clients/venice-thin-client/src/main/java/com/linkedin/venice/client/store/AbstractAvroStoreClient.java

```diff
@@ -657,59 +634,96 @@ public void compute(
       return;
     }
 
-    Optional<ClientStats> clientStats = Optional.empty();
-    if (callback instanceof TrackingStreamingCallback) {
-      clientStats = Optional.of(((TrackingStreamingCallback<K, V>) callback).getStats());
-    }
-
     List<K> keyList = new ArrayList<>(keys);
-    long preRequestSerializationNanos = System.nanoTime();
-    byte[] serializedComputeRequest = serializeComputeRequest(computeRequestWrapper, keyList);
-    clientStats.ifPresent(
-        stats -> stats.recordRequestSerializationTime(LatencyUtils.getLatencyInMS(preRequestSerializationNanos)));
 
-    int schemaId = getSchemaReader().getValueSchemaId(computeRequestWrapper.getValueSchema());
-    Map<String, String> headerMap = new HashMap<>(
-        (computeRequestWrapper.getComputeRequestVersion() == COMPUTE_REQUEST_VERSION_V2)
+    Optional<ClientStats> stats = getClientStats(callback);
+    ClientComputeStreamingCallback clientComputeCallback = new ClientComputeStreamingCallback<K, GenericRecord>() {
+      Map<String, Object> sharedContext = new HashMap<>();
+
+      @Override
+      public void onRawRecordReceived(K key, GenericRecord value) {
+        if (value != null) {
+          value = ComputeUtils.computeResult(
+              computeRequest.getComputeRequestVersion(),
+              computeRequest.getOperations(),
+              sharedContext,
+              value,
+              resultSchema);
+          stats.ifPresent(s -> s.recordMultiGetFallback(1));
+        }
+        onRecordReceived(key, value);
+      }
+
+      @Override
+      public void onRecordReceived(K key, GenericRecord value) {
+        callback.onRecordReceived(
+            key,
+            value != null ? new ComputeGenericRecord(value, computeRequest.getValueSchema()) : null);
+      }
+
+      @Override
+      public void onRemoteComputeStateChange(boolean enabled) {
+        remoteComputationAllowed.set(enabled);
+      }
+
+      @Override
+      public void onCompletion(Optional<Exception> exception) {
+        callback.onCompletion(exception);
+      }
+    };
+
+    TransportClientStreamingCallback decoder = new ClientComputeRecordStreamDecoder<K, GenericRecord>(
+        keyList,
+        clientComputeCallback,
+        callback,
+        getDeserializationExecutor(),
+        streamingFooterRecordDeserializer,
+        () -> getComputeResultRecordDeserializer(resultSchema),
+        schemaId -> (RecordDeserializer<GenericRecord>) getDataRecordDeserializer(schemaId),
+        this::decompressRecord);
+
+    if (clientConfig.isRemoteComputationOnly() || remoteComputationAllowed.get()) {
```

Comment:
In certain cases, the latency difference between remote and client computation can be dramatic. I've added `isRemoteComputationOnly` so that use-cases that enabled remote-computation can detect service operator mistakes quickly.